lib/route/cls/flower.c

```diff
@@ -0,0 +1,779 @@
+#include <netlink-private/route/tc-api.h>
+#include <netlink/route/classifier.h>
+#include <netlink/route/action.h>
+#include <netlink/route/cls/flower.h>
+
+
+/** @cond SKIP */
+#define FLOWER_ATTR_FLAGS           1 << 0x0
+#define FLOWER_ATTR_ACTION          1 << 0x1
+#define FLOWER_ATTR_POLICE          1 << 0x2
+#define FLOWER_ATTR_VLAN_ID         1 << 0x3
+#define FLOWER_ATTR_VLAN_PRIO       1 << 0x4
+#define FLOWER_ATTR_VLAN_ETH_TYPE   1 << 0x5
+#define FLOWER_ATTR_DST_MAC         1 << 0x6
+#define FLOWER_ATTR_DST_MAC_MASK    1 << 0x7
+#define FLOWER_ATTR_SRC_MAC         1 << 0x8
+#define FLOWER_ATTR_SRC_MAC_MASK    1 << 0x9
+#define FLOWER_ATTR_IP_DSCP         1 << 0xa
+#define FLOWER_ATTR_IP_DSCP_MASK    1 << 0xb
+#define FLOWER_ATTR_PROTO           1 << 0xc
+/** @endcond */
+
+#define FLOWER_DSCP_MAX             0xe0
+#define FLOWER_DSCP_MASK_MAX        0xe0
+#define FLOWER_VID_MAX              4095
+#define FLOWER_VLAN_PRIO_MAX        7
+
+static struct nla_policy flower_policy[TCA_FLOWER_MAX + 1] = {
+        [TCA_FLOWER_KEY_ETH_TYPE]	= { .type = NLA_U16 },
+        [TCA_FLOWER_KEY_ETH_DST]	= { .maxlen = ETH_ALEN },
+	[TCA_FLOWER_KEY_ETH_DST_MASK]	= { .maxlen = ETH_ALEN },
+	[TCA_FLOWER_KEY_ETH_SRC]	= { .maxlen = ETH_ALEN },
+	[TCA_FLOWER_KEY_ETH_SRC_MASK]	= { .maxlen = ETH_ALEN },
+        [TCA_FLOWER_KEY_VLAN_ID]	= { .type = NLA_U16 },
+	[TCA_FLOWER_KEY_VLAN_PRIO]	= { .type = NLA_U8 },
+        [TCA_FLOWER_KEY_IP_TOS]		= { .type = NLA_U8 },
+	[TCA_FLOWER_KEY_IP_TOS_MASK]	= { .type = NLA_U8 },
+};
+
+static int flower_msg_parser(struct rtnl_tc *tc, void *data)
+{
+        struct rtnl_flower *f = data;
+	struct nlattr *tb[TCA_FLOWER_MAX + 1];
+	int err;
+
+	err = tca_parse(tb, TCA_FLOWER_MAX, tc, flower_policy);
+	if (err < 0)
+		return err;
+
+	if (tb[TCA_FLOWER_FLAGS]) {
+		f->cf_flags = nla_get_u32(tb[TCA_FLOWER_FLAGS]);
+		f->cf_mask |= FLOWER_ATTR_FLAGS;
+	}
+
+	if (tb[TCA_FLOWER_ACT]) {
+		err = rtnl_act_parse(&f->cf_act, tb[TCA_FLOWER_ACT]);
+		if (err)
+			return err;
+
+		f->cf_mask |= FLOWER_ATTR_ACTION;
+	}
+
+	if (tb[TCA_FLOWER_POLICE]) {
+		f->cf_police = nl_data_alloc_attr(tb[TCA_FLOWER_POLICE]);
+		if (!f->cf_police)
+			return -NLE_NOMEM;
+
+		f->cf_mask |= FLOWER_ATTR_POLICE;
+	}
+
+	if (tb[TCA_FLOWER_KEY_ETH_TYPE]) {
+	        f->cf_proto = nla_get_u16(tb[TCA_FLOWER_KEY_ETH_TYPE]);
+		f->cf_mask |= FLOWER_ATTR_PROTO;
+	}
+
+	if (tb[TCA_FLOWER_KEY_VLAN_ID]) {
+	        f->cf_vlan_id = nla_get_u16(tb[TCA_FLOWER_KEY_VLAN_ID]);
+		f->cf_mask |= FLOWER_ATTR_VLAN_ID;
+	}
+
+	if (tb[TCA_FLOWER_KEY_VLAN_PRIO]) {
+	        f->cf_vlan_prio = nla_get_u16(tb[TCA_FLOWER_KEY_VLAN_PRIO]);
+		f->cf_mask |= FLOWER_ATTR_VLAN_PRIO;
+	}
+
+	if (tb[TCA_FLOWER_KEY_VLAN_ETH_TYPE]) {
+	        f->cf_vlan_ethtype = nla_get_u16(tb[TCA_FLOWER_KEY_VLAN_ETH_TYPE]);
+		f->cf_mask |= FLOWER_ATTR_VLAN_ETH_TYPE;
+	}
+
+	if (tb[TCA_FLOWER_KEY_ETH_DST]) {
+	        f->cf_dst_mac = nl_data_alloc_attr(tb[TCA_FLOWER_KEY_ETH_DST]);
```

Comment:
the MAC addresses are all fixed size of `ETH_ALEN` bytes. Maybe better to embed these fields as as arrays:
```
uint8_t cf_dst_mac[ETH_ALEN];
```
Compare to `struct rtnl_link`'s `l_addr`, which is of type `struct nl_addr`. That would be a possibility, however, here the length is always fixed, so embedding the array in the struct seems preferable to me.

Note also API like `rtnl_flower_get_dst_mac()` where the length `ETH_ALEN` is fixed in the API. If you'd ever believe the length could be any different, then this API would need to change so that the length of the MAC address can be specified/retrieved.

Reply:
Sure, I think making mac/mask fields as arrays, as you suggested, is pretty good.